Hash full Lynx file content

rzumer opened the pull request:
## Description

This PR mirrors https://github.com/RetroAchievements/RALibretro/pull/58/ and skips hashing completely in the absence of a valid header. Headerless files should simply fall back to `GENERIC_MD5` as is done for NES and SNES.

1. Detect the Lynx header and skip (fall back to `GENERIC_MD5`) if it does not exist.
2. Hash full file contents without the unnecessary 512KB limit.

## Reviewers

@leiradel

RobLoach commented:
Thank you for doing this! Will be able to test today.

Would love to see this used for the scanner too. NES and Lynx games often don't show up for users because of the header.

leiradel commented:
Thanks for the PR. Could you change

```c
coro->count  = coro->len - coro->offset;
```

to

```c
coro->count  = coro->len - LYNX_HEADER_LEN;
```

?

I believe the intent is more clear that way.

Also please for the sake of coding style, rename `LYNX_HEADER_LEN` to be in small caps, or make it a macro.

leiradel commented:
> Would love to see this used for the scanner too. NES and Lynx games often don't show up for users because of the header.

With the recent PRs regarding ROM hashing for RetroAchievements, it should be easy to update the scanner to also skip NES and Lynx headers.

RobLoach commented on cheevos/cheevos.c:
```diff
@@ -2955,16 +2955,21 @@ static int cheevos_iterate(coro_t *coro)
+      const int LYNX_HEADER_LEN = 0x40;
+
+      if (coro->len <= LYNX_HEADER_LEN ||
+        memcmp("LYNX", (void *)coro->data, 5) != 0)
```
`5` is a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming))?

leiradel replied:
Literals in C are guaranteed by the standard to end with a null terminator. The Lynx ROM header also has a 0 byte after `LYNX`, so `memcmp` is comparing the four letters and the null terminator in one go.

RobLoach replied:
Sounds good, thanks :+1:

rzumer commented:
> Thanks for the PR. Could you change
> 
> ```c
> coro->count  = coro->len - coro->offset;
> ```
> 
> to
> 
> ```c
> coro->count  = coro->len - LYNX_HEADER_LEN;
> ```
> 
> ?
> 
> I believe the intent is more clear that way.

OK, done.

> Also please for the sake of coding style, rename `LYNX_HEADER_LEN` to be in small caps, or make it a macro.

Done, moved at the top for C89 compliance and renamed `SNES_HEADER_LEN` to match.

leiradel commented:
> Done, moved at the top for C89 compliance and renamed `SNES_HEADER_LEN` to match.

Nice, I always forget about C89 compliance.

i30817 commented:
Does this means that snes and lynx and nes and pc-engine turbografx hashes in libretro-database should not include the header now?

rzumer commented:
> Does this means that snes and lynx and nes and pc-engine turbografx hashes in libretro-database should not include the header now? Just wondering for my hack PRs to there.

As far as I know this PR *only* affects the RetroAchievements functionality, so any header skipping logic should be duplicated/moved elsewhere in order to be able to take advantage of it outside of the achievement system.

> Also is that check if the rom size is not a multiple of the header size enough? Why shouldn't some headerless game coincidentally be a multiple and be mis-recognized as 'headered'?

I don't think such a file exists currently, but I suppose someone could write homebrew that ends up with such a signature. For our purposes that is not a concern, though it might be for yours. A false positive with the right extension, matching the standard header signature and with the correct size multiple will send a request to the RetroAchievements server, and if no match is found, the system will end up falling back to the generic MD5 (full file content) hashing method anyway.

i30817 commented:
So this means that the database the cheevos use to id games doesn't have the same checksums as librertro (in so far as libretro includes headers for those headered consoles)?

leiradel commented:
> So this means that the database the cheevos use to id games doesn't have the same checksums as libreretro-database (in so far as libretro-database includes headers for those headered consoles, if the dump has a header)? Is it using another, online database instead of the local rdbs?

That's correct.

leiradel commented:
> Is this PR fit to be merged?

I believe so, yes.

RobLoach commented:
With this change, I am no longer able to see Atari - Lynx Achievements with Handy. This test is with [Robotron 2084](https://retroachievements.org/game/12858)...

## Master

![master](https://user-images.githubusercontent.com/25086/51259868-0362f700-197b-11e9-9752-8d46c17dd211.png)

## lynx-cheevos-fix

![screenshot at 2019-01-16 10-40-07](https://user-images.githubusercontent.com/25086/51259938-28f00080-197b-11e9-9c40-2778385a890e.png)

leiradel commented:
> With this change, I am no longer able to see Atari - Lynx Achievements with Handy. This test is with [Robotron 2084](https://retroachievements.org/game/12858) `7117e9f7`...

Can you post a verbose log from RetroArch while opening Robotron 2084? Thanks.

RobLoach commented:
> Can you post a verbose log from RetroArch while opening Robotron 2084?

https://gist.github.com/RobLoach/ae7360edb6ed8426341e6bcf3ddf85d8

leiradel commented:
Hm, there aren't `[CHEEVOS]` entries in the log. Can you filter the log and post the result?

```
$ retroarch ... 2>&1 |grep CHEEVOS
```

Thanks

leiradel commented:
Sorry, I forgot that you have to define `CHEEVOS_VERBOSE` and rebuild to get log messages from cheevos: https://github.com/libretro/RetroArch/blob/ac14ea6d019cf6d2e5f4d9457c599c0575ad3460/cheevos-new/util.h#L28

That made it so much more difficult to give support for issues with achievements :/

RobLoach commented:
After switching it to `#define CHEEVOS_VERBOSE`....
https://gist.github.com/RobLoach/79e11b54b08df81ff0749e0d13477c8b

leiradel commented:
Shouldn't we have a fallback then? Or maybe someone could re-hash all Atari Lynx games and update the database, and we revert this PR until that is done?

leiradel commented:
Woot that was fast. All is good now.

rzumer commented:
@meleu OK, thank you.
